### PR TITLE
added optitrack data collection

### PR DESCRIPTION
**Pull request by dhruvpatelaj**

### CODE REVIEW

File: launch/test.launch [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    <rosparam subst_value="true">
+      server: $(arg server)
+      port: 3883
+      update_frequency: 100.0
+      frame_id: optitrack_world
+      # Use the VRPN server's time, or the client's ROS time.
+      use_server_time: false
+      broadcast_tf: true
+      # Must either specify refresh frequency > 0.0, or a list of trackers to create
+      refresh_tracker_frequency: 1.0
```
Comment (peasant98) [inline launch/test.launch]: Why is this called `test.launch`? I would suggest a more descriptive name here.

**Review by peasant98 — COMMENTED** [review]
Plz add a description to your pr

File: scripts/optitrack_subscriber.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+def save(pickle_data):
+    ros_robotic_skin_path = rospkg.RosPack().get_path('ros_robotic_skin')
+    filepath = os.path.join(ros_robotic_skin_path, 'data/optitrack_data.pickle')
+
+    with open(filepath, 'ab') as f:
+        pickle.dump(pickle_data, f)
+
+
+def callback(data):
+    # rospy.loginfo(rospy.get_caller_id() + 'I heard %s ', data)
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: needed?

File: launch/test.launch (hunk 1 above)
Comment (dhruvpatelaj) [reply launch/test.launch]: I'll fix that later. -BJ

File: scripts/optitrack_subscriber.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import os
+import pickle
+import rospy
+import rospkg
+from geometry_msgs.msg import PoseStamped
+
+
+def save(pickle_data):
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: even though some functions in the codebase may not have it, I highly recommend docstrings. Check out https://www.python.org/dev/peps/pep-0257/. They succinctly describe what the function is going.

File: scripts/optitrack_subscriber.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    ros_robotic_skin_path = rospkg.RosPack().get_path('ros_robotic_skin')
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: this doesn't need to be called every time `save` is called, I would recommend calling it outside of this function - only called once.

File: scripts/optitrack_subscriber.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    filepath = os.path.join(ros_robotic_skin_path, 'data/optitrack_data.pickle')
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: Same as the previous line in `data/optitrack_data.pickle` doesn't change

File: scripts/optitrack_subscriber.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    topic_name = "/vrpn_client_node/RigidBody01/pose"
+    topic_type = PoseStamped
+    rospy.Subscriber(topic_name, topic_type, callback)
+
+    # spin() simply keeps python from exiting until this node is stopped
+    rospy.sleep(3)
+
+
+if __name__ == '__main__':
+    ros_robotic_skin_path = rospkg.RosPack().get_path('ros_robotic_skin')
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: I'd put this at the top of the file as well

File: scripts/optitrack_subscriber.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def listener():
+    rospy.init_node('listener', anonymous=True)
+
+    topic_name = "/vrpn_client_node/RigidBody01/pose"
+    topic_type = PoseStamped
+    rospy.Subscriber(topic_name, topic_type, callback)
+
+    # spin() simply keeps python from exiting until this node is stopped
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: where is `spin()`?

File: scripts/optitrack_subscriber.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+def callback(data):
+    # rospy.loginfo(rospy.get_caller_id() + 'I heard %s ', data)
+    save(data)
+
+
+def listener():
+    rospy.init_node('listener', anonymous=True)
+
+    topic_name = "/vrpn_client_node/RigidBody01/pose"
+    topic_type = PoseStamped
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: these variables doesn't need to be declared if they are used just once

File: scripts/optitrack_subscriber.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    rospy.Subscriber(topic_name, topic_type, callback)
+
+    # spin() simply keeps python from exiting until this node is stopped
+    rospy.sleep(3)
+
+
+if __name__ == '__main__':
+    ros_robotic_skin_path = rospkg.RosPack().get_path('ros_robotic_skin')
+    filepath = os.path.join(ros_robotic_skin_path, 'data/optitrack_data.pickle')
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: I'm trying to remember ... I do think if you write to a pickle again, you will overwrite the contents - so no need for this.

File: scripts/optitrack_subscriber.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+def save(pickle_data):
+    
+    ros_robotic_skin_path = rospkg.RosPack().get_path('ros_robotic_skin')
+    filepath = os.path.join(ros_robotic_skin_path, 'data/optitrack_data.pickle')
+
+    with open(filepath, 'ab') as f:
+        pickle.dump(pickle_data, f)
+
+def callback(data):
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: I am confused here. Does this save to a pickle file everytime the callback is called?

File: scripts/optitrack_subscriber.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    #rospy.loginfo(rospy.get_caller_id() + 'I heard %s ', data)
+    save(data)
+
+def listener():
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: Plz use a more descriptive function name

File: scripts/optitrack_subscriber.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+if __name__ == '__main__':
+    ros_robotic_skin_path = rospkg.RosPack().get_path('ros_robotic_skin')
+    filepath = os.path.join(ros_robotic_skin_path, 'data/optitrack_data.pickle')
+    if os.path.exists(filepath):
+        os.remove(filepath)
+    
+    listener()
+   
+    
+    # with open(filepath, 'rb') as f:
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: delete if not used

File: scripts/optitrack_subscriber.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        pickle.dump(pickle_data, f)
+
+def callback(data):
+    #rospy.loginfo(rospy.get_caller_id() + 'I heard %s ', data)
+    save(data)
+
+def listener():
+
+
+    rospy.init_node('listener', anonymous=True)
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: consider using `optitrack_listener`?

File: scripts/optitrack_subscriber.py (hunk 10 above)
Comment (peasant98) [reply scripts/optitrack_subscriber.py]: If the callback is called at 20hz, it is saving one message to a file 20 times a second. Instead, I would recommend collecting the data, then saving to a file (as a one time thing), which should be faster.

File: scripts/optitrack_subscriber.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+    topic_name = "/vrpn_client_node/RigidBody01/pose"
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: use a constant for this?

File: scripts/optitrack_subscriber.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
+import os
+import pickle
+import rospy
+import rospkg
+from geometry_msgs.msg import PoseStamped
+
+def save(pickle_data):
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: I would recommend a function name such as `save_to_pickle_file` - `save` can mean a lot of things

**Review by peasant98 — CHANGES_REQUESTED** [review]

File: launch/test.launch (hunk 1 above)
Comment (peasant98) [reply launch/test.launch]: resolving, as this is fixed.

File: scripts/optitrack_subscriber.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+import pickle
+import rospy
+import rospkg
+from geometry_msgs.msg import PoseStamped
+
+ROS_PATH = rospkg.RosPack().get_path('ros_robotic_skin')
+FILEPATH = os.path.join(ROS_PATH, 'data/optitrack_data.pickle')
+
+
+class poseData:
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: I recommend using OptriTrackPoseData as the name.

File: scripts/optitrack_subscriber.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+if __name__ == '__main__':
+
+    # accesses command line arguments through rospy
+    args = rospy.myargv(argv=sys.argv)
+
+    if len(args) != 2:
+        raise ValueError("RigidBody not provided!")
+
+    topic = ["/vrpn_client_node/"+args[1]+"/pose", 'geometry_msgs/PoseStamped']
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: nit: in python2, you can use .format() (https://www.geeksforgeeks.org/python-format-function/) for strings

File: scripts/optitrack_subscriber.py [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    if topic not in rospy.get_published_topics():
+        raise ValueError("RigidBody does not exist!")
+
+    newSave = poseData()
```
Comment (peasant98) [inline scripts/optitrack_subscriber.py]: newSave is a bit of a cofusing name - maybe use OptriTrackPoseDataSaver?

File: scripts/optitrack_subscriber.py (hunk 16 above)
Comment (peasant98) [reply scripts/optitrack_subscriber.py]: or OptriTrackPoseDataSaver (that one is better imo)

Comment (dhruvpatelaj) [reply scripts/optitrack_subscriber.py]: changed to OptiTrackPoseDataSaver

File: scripts/optitrack_subscriber.py (hunk 17 above)
Comment (dhruvpatelaj) [reply scripts/optitrack_subscriber.py]: changed

File: scripts/optitrack_subscriber.py (hunk 18 above)
Comment (dhruvpatelaj) [reply scripts/optitrack_subscriber.py]: changed to optitrackPoseData

**Review by peasant98 — APPROVED** [review]

File: scripts/optitrack_subscriber.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    Class for collecting all pose data for the objects in optitrack
+    """
+    def __init__(self):
+        """
+        Initializes the list that stores all data
+        """
+        self.pose_data = []
+
+    def save_to_pickle(self, pickle_data):
```
Comment (watakandai) [inline scripts/optitrack_subscriber.py]: `append_pose` might be a better name

File: scripts/optitrack_subscriber.py [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    `obj`: `String`
+        The object the optitrack listener will track
+    """
+
+    # initializes the subscriber
+    rospy.init_node('optitrack_listener', anonymous=True)
+    rospy.Subscriber(topic[0], PoseStamped, poseDataClass.save_to_pickle)
+
+    # saves 3 seconds of data
+    rospy.sleep(3)
```
Comment (watakandai) [inline scripts/optitrack_subscriber.py]: make it an argument

File: scripts/optitrack_subscriber.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+        """
+        Adds all optitrack data to a list
+        Arguments
+        ----------
+        `picke_data`: `PoseStamped`
+            The data to be added to the list
+        """
+        self.pose_data.append(pickle_data)
+
+    def writePickleFile(self):
```
Comment (watakandai) [inline scripts/optitrack_subscriber.py]: This is a very small thing but we don't use camel cases in python.
Instead, connect small cases with an underscore like `write_pickle_file()`

**Review by watakandai — CHANGES_REQUESTED** [review]

File: scripts/optitrack_subscriber.py (hunk 21 above)
Comment (peasant98) [reply scripts/optitrack_subscriber.py]: ah good catch

**Comment by peasant98** [comment]
cool. lgtm.

**Review by watakandai — APPROVED** [review]

**Pull request merged**